cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ArtifactHttpHandler.java

```diff
@@ -0,0 +1,422 @@
+      return null;
+    }
+
+    // find out if this artifact extends other artifacts. If so, there will be a header like
+    // 'Artifact-Extends: <name>[<lowerversion>,<upperversion>]/<name>[<lowerversion>,<upperversion>]:
+    // for example: 'Artifact-Extends: etl-batch[1.0.0,2.0.0]/etl-realtime[1.0.0:3.0.0]
+    final Set<ArtifactRange> parentArtifacts = Sets.newHashSet();
+    if (parentArtifactsStr != null) {
+      for (String parent : Splitter.on('/').split(parentArtifactsStr)) {
+        try {
+          parentArtifacts.add(ArtifactRange.parse(namespace, parent));
+        } catch (InvalidArtifactRangeException e) {
+          responder.sendString(HttpResponseStatus.BAD_REQUEST, e.getMessage());
+          return null;
+        }
+      }
+    }
+
+    try {
+      // copy the artifact contents to local tmp directory
+      final File destination = File.createTempFile("artifact-", ".jar", tmpDir);
+
+      return new AbstractBodyConsumer(destination) {
+
+        @Override
+        protected void onFinish(HttpResponder responder, File uploadedFile) {
+          try {
+            // add the artifact to the repo
+            artifactRepository.addArtifact(artifactId, uploadedFile, parentArtifacts);
+            responder.sendString(HttpResponseStatus.OK, "Artifact added successfully");
+          } catch (InvalidArtifactException e) {
+            responder.sendString(HttpResponseStatus.BAD_REQUEST, e.getMessage());
+          } catch (ArtifactRangeNotFoundException e) {
+            responder.sendString(HttpResponseStatus.NOT_FOUND, e.getMessage());
+          } catch (ArtifactAlreadyExistsException e) {
+            responder.sendString(HttpResponseStatus.CONFLICT, "Artifact " + artifactId + " already exists.");
+          } catch (WriteConflictException e) {
+            responder.sendString(HttpResponseStatus.INTERNAL_SERVER_ERROR,
+              "Conflict while writing artifact, please try again");
+          } catch (IOException e) {
+            LOG.error("Exception while trying to write artifact {}.", artifactId, e);
+            responder.sendString(HttpResponseStatus.INTERNAL_SERVER_ERROR,
+              "Error performing IO while writing artifact");
+          }
+        }
+      };
+    } catch (IOException e) {
+      LOG.error("Exception creating temp file to place artifact {} contents", artifactId, e);
+      responder.sendString(HttpResponseStatus.INTERNAL_SERVER_ERROR, "Server error creating temp file for artifact.");
+      return null;
+    }
+  }
+
+  // check that the namespace exists, writing a response if it doesn't and returning null.
+  // also checks if the request is only supposed to include system artifacts, and returning
+  // the system namespace if so.
+  private Id.Namespace validateAndGetNamespace(String namespaceId, boolean isSystem, HttpResponder responder) {
+    Id.Namespace namespace = Id.Namespace.from(namespaceId);
+    if (!namespaceAdmin.hasNamespace(namespace)) {
```

Comment:
it will be false until we actually create the namespace, which is not in this PR.